### PR TITLE
executor: correctly handle panic for hashjoin build phase

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/14027

### What is changed and how it works?
handle the panic when OOM during build a hash table

### Check List 

Tests 

 - No code

Code changes

 - N/A

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release branch
3.0.5

Release note

 - correctly output OOM errors when building a hash table

### CODE REVIEW

File: executor/join.go [hunk 1]
```diff
@@ -647,6 +647,12 @@ func (e *HashJoinExec) handleFetchAndBuildHashTablePanic(r interface{}) {
 	close(e.buildFinished)
 }
 
+func (e *HashJoinExec) handleFetchBuildSidePanic(r interface{}) {
```
Comment (SunRunAway) [inline executor/join.go]: Where to use this method?

File: executor/join.go [hunk 2]
```diff
@@ -16,6 +16,7 @@ package executor
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/failpoint"
```
Comment (zz-jason) [inline executor/join.go]: please reorg the import packages.

File: executor/join_test.go [hunk 3]
```diff
@@ -16,6 +16,7 @@ package executor_test
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/failpoint"
```
Comment (jackysp) [inline executor/join_test.go]: Please move it to the 3rd party libs.

File: executor/join.go (hunk 2 above)
Comment (fzhedu) [reply executor/join.go]: done

File: executor/join_test.go (hunk 3 above)
Comment (fzhedu) [reply executor/join_test.go]: done

File: executor/join.go [hunk 4]
```diff
@@ -244,6 +246,11 @@ func (e *HashJoinExec) fetchBuildSideRows(ctx context.Context, chkCh chan<- *chu
 			e.buildFinished <- errors.Trace(err)
 			return
 		}
+		failpoint.Inject("errorFetchBuildSideRowsMockOOMPanic", func(val failpoint.Value) {
```
Comment (lonng) [inline executor/join.go]: `failpoint.Inject("errorFetchBuildSideRowsMockOOMPanic", nil)` and `c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/errorFetchBuildSideRowsMockOOMPanic", `panic`), IsNil)`

File: executor/join_test.go [hunk 5]
```diff
@@ -78,6 +79,20 @@ func (s *testSuite) TestJoinInDisk(c *C) {
 	result.Check(testkit.Rows("2 2 2 3"))
 }
 
+func (s *testSuite) TestOOMPanicInHashJoinWhenFetchBuildRows(c *C) {
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/errorFetchBuildSideRowsMockOOMPanic", `return(true)`), IsNil)
+	defer func() {
+		c.Assert(failpoint.Disable("github.com/pingcap/tidb/executor/errorFetchBuildSideRowsMockOOMPanic"), IsNil)
+	}()
```
Comment (lonng) [inline executor/join_test.go]: Prefer to:
```
fpName := "github.com/pingcap/tidb/executor/errorFetchBuildSideRowsMockOOMPanic"
c.Assert(failpoint.Enable(fpName, `panic`), IsNil)
	defer func() {
		c.Assert(failpoint.Disable(fpName), IsNil)
	}()
```

File: executor/join.go [hunk 6]
```diff
@@ -232,6 +233,7 @@ var buildSideResultLabel fmt.Stringer = stringutil.StringerStr("hashJoin.buildSi
 // fetchBuildSideRows fetches all rows from build side executor, and append them
 // to e.buildSideResult.
 func (e *HashJoinExec) fetchBuildSideRows(ctx context.Context, chkCh chan<- *chunk.Chunk, doneCh <-chan struct{}) {
+
```
Comment (zz-jason) [inline executor/join.go]: please remove the empty line.

File: executor/join.go [hunk 7]
```diff
@@ -643,30 +658,32 @@ func (e *HashJoinExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
...
-	go util.WithRecovery(func() { e.fetchBuildSideRows(ctx, buildSideResultCh, doneCh) }, nil)
+	go util.WithRecovery(func() { e.fetchBuildSideRows(ctx, buildSideResultCh, doneCh) }, e.handleFetchBuildSidePanic)
 
 	// TODO: Parallel build hash table. Currently not support because `rowHashMap` is not thread-safe.
 	err := e.buildHashTableForList(buildSideResultCh)
 	if err != nil {
-		e.buildFinished <- errors.Trace(err)
+		e.buildHashtableFinished <- err
 		close(doneCh)
 	}
```
Comment (SunRunAway) [inline executor/join.go]: We can create the temparory variable `fetchBuildSideFinished` (not nessacerray to be a entry of `HashJoinExec`) before `go e.fetchBuildSideRows` and check the result of this channel immediately after `buildHashTableForList` finished.

Comment (XuHuaiyu) [reply executor/join.go]: But why do we need this variable?

**Pull request merged**